File: examples/python/conformance/server_app.py
from __future__ import annotations

import os
import struct
from dataclasses import dataclass
from typing import Any

import uvicorn
from fastapi import FastAPI, Request
from starlette.responses import JSONResponse

from api_blueprint_example_server.alt.routes.alt.conflict.gen_types import (
    DefaultQuery as AltDefaultQuery,
)
from api_blueprint_example_server.alt.routes.alt.conflict.gen_types import (
    DefaultResponse as AltDefaultResponse,
)
from api_blueprint_example_server.alt.routes.alt.conflict.gen_types import (
    KeywordEnum as AltKeywordEnum,
)
from api_blueprint_example_server.alt.transports.http.server import create_router as create_alt_router
from api_blueprint_example_server.api.routes.api.binary.gen_types import (
    AuditPacketQuery,
    AuditPacketResponse,
    PacketQuery,
    PacketResponse,
)
from api_blueprint_example_server.api.routes.api.conflict.gen_types import (
    DefaultQuery,
    DefaultResponse,
    KeywordEnum,
)
from api_blueprint_example_server.api.routes.api.demo.gen_types import (
    ANONDeleteAnonList,
    ANONFunc1putAnonKv,
    AbcQuery,
    AbcResponse,
    ApiDemoA,
    ApiDemoMap,
    ApiDemoSubA,
    AssistantCancel,
    AssistantDelta,
    AssistantInput,
    AssistantServerMessageVariants,
    AssistantSessionClose,
    AssistantSessionOpen,
    ColorEnum,
    DeleteQuery,
    DeleteResponse,
    ErrorDemoQuery,
    ErrorDemoResponse,
    FormSubmitForm,
    FormSubmitResponse,
    PostDeprecatedJSON,
    PostDeprecatedResponse,
    PutDemoJSON,
    PutDemoQuery,
    PutDemoResponse,
    RawResponse,
    StatusEnum,
    SweepEventsClose,
    SweepEventsOpen,
    SweepState,
    SweepStreamMessageVariants,
    TestPostJSON,
    TestPostResponse,
)
from api_blueprint_example_server.api.routes.api.gen_types import (
    HelloChannelClose,
    HelloChannelMessage,
    HelloChannelMsgTypeEnum,
)
from api_blueprint_example_server.api.routes.api.hello.gen_types import (
    AbcQuery as HelloAbcQuery,
)
from api_blueprint_example_server.api.routes.api.hello.gen_types import ApiHelloMap, HelloWayQuery, MapEnum
from api_blueprint_example_server.api.runtime.errors import ApiError, ApiErrorPayload, ApiToastPayload
from api_blueprint_example_server.api.transports.http.server import create_router as create_api_router
from api_blueprint_example_server.static.routes.static.gen_types import DocJsonResponse, DochahaResponse
from api_blueprint_example_server.static.transports.http.server import create_router as create_static_router


app = FastAPI()


@app.middleware("http")
async def require_demo_header(request: Request, call_next):
    if request.url.path == "/api/demo/abc" and request.headers.get("x-token") != "conformance-token":
        return JSONResponse({"detail": "missing conformance token"}, status_code=418)
    return await call_next(request)


class ApiService:
    async def hello_channel(self, channel: Any = None) -> None:
        first = await channel.receive()
        if first is not None:
            await channel.send(
                HelloChannelMessage(type=HelloChannelMsgTypeEnum.PONG, data={"source": "python"})
            )
        await channel.close(HelloChannelClose(code=1000, reason="single channel complete"))


class DemoService:
    async def abc(self, query: AbcQuery) -> AbcResponse:
        return demo_model("header-ok")

    async def test_post(self, json: TestPostJSON) -> TestPostResponse:
        req1 = json.req1
        req2 = 0 if json.req2 is None else json.req2
        return TestPostResponse(
            list=["test_post", req1],
            map={"req2": ApiDemoMap(haha=req2)},
        )

    async def form_submit(self, form: FormSubmitForm) -> FormSubmitResponse:
        return FormSubmitResponse(
            summary=form.title,
            count=0 if form.count is None else form.count,
            enabled=False if form.enabled is None else form.enabled,
        )

    async def put_demo(
        self,
        query: PutDemoQuery,
        json: PutDemoJSON,
    ) -> PutDemoResponse:
        arg1 = "" if query.arg1 is None else query.arg1
        arg2 = 0.0 if query.arg2 is None else query.arg2
        req1 = json.req1
        req2 = 0 if json.req2 is None else json.req2
        return PutDemoResponse(
            list=[arg1, req1],
            anon_kv=ANONFunc1putAnonKv(kv1=req2, kv2=[arg2, float(req2)]),
        )

    async def delete(self, query: DeleteQuery) -> DeleteResponse:
        return DeleteResponse(
            list=["" if query.arg1 is None else query.arg1],
            anon_list=[ANONDeleteAnonList(kv1=7, kv2=["xml"])],
        )

    async def post_deprecated(self, json: PostDeprecatedJSON) -> PostDeprecatedResponse:
        return PostDeprecatedResponse(list=[json.req1])

    async def raw(self) -> RawResponse:
        return RawResponse(list=["raw"], list2={1: [demo_model("raw")]})

    async def map_model(self) -> dict[int, ApiDemoMap]:
        return {1: ApiDemoMap(haha=101)}

    async def error_demo(self, query: ErrorDemoQuery) -> ErrorDemoResponse:
        mode = "ok" if query.mode is None else query.mode
        if mode == "rate_limit":
            raise ApiError(
                ApiErrorPayload(
                    id="DemoErr.RATE_LIMITED",
                    group="",
                    key="",
                    code=42901,
                    message="",
                    toast=ApiToastPayload(
                        key="demo.rate_limited",
                        level="warning",
                        default="请求过于频繁，请稍后再试",
                        text="请等待 30 秒后重试",
                    ),
                )
            )
        if mode == "unknown":
            raise ApiError(
                ApiErrorPayload(
                    id="",
                    group="",
                    key="",
                    code=70001,
                    message="example undefined business error",
                    toast=ApiToastPayload(level="error"),
                )
            )
        return ErrorDemoResponse(status="ok")

    async def sweep_events(
        self,
        open_data: SweepEventsOpen,
        stream: Any = None,
    ) -> None:
        run_id = open_data.run_id
        await stream.send(SweepStreamMessageVariants.state(SweepState(status=f"python sweep {run_id}")))
        await stream.close(SweepEventsClose(code=1000, reason="example stream complete"))

    async def assistant_session(
        self,
        open_data: AssistantSessionOpen,
        channel: Any = None,
    ) -> None:
        session_id = open_data.session_id
        first = await channel.receive()
        if first.type == "input":
            payload = AssistantInput.from_value(first.data)
            await channel.send(
                AssistantServerMessageVariants.delta(AssistantDelta(text=f"{session_id}:{payload.text}"))
            )
        elif first.type == "cancel":
            payload = AssistantCancel.from_value(first.data)
            await channel.close(AssistantSessionClose(code=1000, reason=payload.reason))
            return
        second = await channel.receive()
        if second.type == "cancel":
            payload = AssistantCancel.from_value(second.data)
            await channel.close(AssistantSessionClose(code=1000, reason=payload.reason))


class BinaryService:
    async def packet(
        self,
        query: PacketQuery,
        binary: bytes | None = None,
    ) -> PacketResponse:
        packet = parse_packet(binary or b"")
        return PacketResponse(
            trace="" if query.trace is None else query.trace,
            version=packet.version,
            item_count=len(packet.item_ids),
            payload=packet.payload,
            score_sum=round(packet.score_sum),
            first_label=packet.first_label,
            item_ids=packet.item_ids,
            checksum=packet.checksum,
        )

    async def audit_packet(
        self,
        query: AuditPacketQuery,
        binary: bytes | None = None,
    ) -> AuditPacketResponse:
        packet = parse_audit_packet(binary or b"")
        return AuditPacketResponse(
            trace="" if query.trace is None else query.trace,
            item_count=packet.item_count,
            checksum=packet.checksum,
        )


class HelloService:
    async def abc(self, query: HelloAbcQuery) -> dict[str, ApiHelloMap]:
        key = "ping" if query.type is None else query.type.value
        return {"hello": ApiHelloMap(haha=1001), key: ApiHelloMap(haha=1)}

    async def map_enum(self) -> dict[MapEnum, ApiHelloMap]:
        return {MapEnum.A: ApiHelloMap(haha=11), MapEnum.B: ApiHelloMap(haha=22)}

    async def list_enum(self) -> list[MapEnum]:
        return [MapEnum.A, MapEnum.B]

    async def string(self) -> str:
        return "hello-string"

    async def uint64(self) -> int:
        return 9007199254740991

    async def string_emun(self) -> MapEnum:
        return MapEnum.A

    async def hello_way(self, query: HelloWayQuery) -> dict[str, Any]:
        return {"echo": None if query.arg1 is None else query.arg1.value}


class StaticService:
    async def doc_json(self) -> DocJsonResponse:
        return DocJsonResponse()

    async def dochaha(self) -> DochahaResponse:
        return DochahaResponse(a="hello world")


class ConflictService:
    async def default(self, query: DefaultQuery) -> DefaultResponse:
        return DefaultResponse(
            default="api-default",
            class_="" if query.class_ is None else query.class_,
            enum=KeywordEnum.DEFAULT,
        )


class AltConflictService:
    async def default(self, query: AltDefaultQuery) -> AltDefaultResponse:
        return AltDefaultResponse(
            default="alt-default",
            class_="" if query.class_ is None else query.class_,
            enum=AltKeywordEnum.CLASS,
        )


app.include_router(
    create_api_router(
        api_service=ApiService(),
        binary_service=BinaryService(),
        conflict_service=ConflictService(),
        demo_service=DemoService(),
        hello_service=HelloService(),
    )
)
app.include_router(create_alt_router(conflict_service=AltConflictService()))
app.include_router(create_static_router(static_service=StaticService()))


@dataclass(frozen=True)
class ParsedPacket:
    version: int
    payload: str
    score_sum: float
    first_label: str
    item_ids: list[int]
    checksum: int


@dataclass(frozen=True)
class ParsedAuditPacket:
    item_count: int
    checksum: int


def demo_model(label: str) -> AbcResponse:
    return AbcResponse(
        bc=label,
        a=1,
        efg=1.5,
        hijk=[1, 2, 3],
        lmnop=[ApiDemoSubA(hello={"a": 1}, amap=[ApiDemoMap(haha=1)])],
        enum_color=ColorEnum.RED,
        enum_status=StatusEnum.RUNNING,
        enum_list=[StatusEnum.PENDING, StatusEnum.RUNNING],
    )


def parse_packet(data: bytes) -> ParsedPacket:
    offset = 0
    magic = data[offset : offset + 4].decode("utf-8")
    offset += 4
    if magic != "ABP1":
        raise ValueError(f"binary magic mismatch: {magic}")
    version, kind = struct.unpack_from("<HH", data, offset)
    offset += 4
    if kind != 1:
        raise ValueError(f"binary kind mismatch: {kind}")
    (flags,) = struct.unpack_from("<I", data, offset)
    offset += 4
    if flags & 1 == 0:
        raise ValueError(f"binary flags missing payload bit: {flags}")
    offset += 3
    short_code = _read_u24(data, offset)
    offset += 3
    if short_code != 0x010203:
        raise ValueError(f"binary short code mismatch: {short_code}")
    offset += 3
    item_count, payload_len, score_count = struct.unpack_from("<HIH", data, offset)
    offset += 8
    item_ids: list[int] = []
    first_label = ""
    for index in range(item_count):
        item_id = struct.unpack_from("<I", data, offset)[0]
        offset += 4
        item_ids.append(item_id)
        offset += 1
        offset += 8
        label_len = data[offset]
        offset += 1
        label = data[offset : offset + label_len].decode("utf-8")
        offset += label_len
        if index == 0:
            first_label = label
    payload = data[offset : offset + payload_len].decode("utf-8")
    offset += payload_len
    score_sum = 0.0
    for _ in range(score_count):
        score_sum += struct.unpack_from("<d", data, offset)[0]
        offset += 8
    checksum = struct.unpack_from("<I", data, offset)[0]
    offset += 4
    if offset != len(data):
        raise ValueError(f"binary packet has trailing bytes: {len(data) - offset}")
    return ParsedPacket(version, payload, score_sum, first_label, item_ids, checksum)


def parse_audit_packet(data: bytes) -> ParsedAuditPacket:
    offset = 0
    (kind,) = struct.unpack_from("<H", data, offset)
    offset += 2
    if kind != 2:
        raise ValueError(f"audit kind mismatch: {kind}")
    (flags,) = struct.unpack_from("<I", data, offset)
    offset += 4
    if flags & 1 == 0:
        raise ValueError(f"audit flags missing items bit: {flags}")
    (item_count,) = struct.unpack_from("<H", data, offset)
    offset += 2
    for _ in range(item_count):
        offset += 4
        offset += 2
    (checksum,) = struct.unpack_from("<I", data, offset)
    offset += 4
    if offset != len(data):
        raise ValueError(f"audit packet has trailing bytes: {len(data) - offset}")
    return ParsedAuditPacket(item_count=item_count, checksum=checksum)


def _read_u24(data: bytes, offset: int) -> int:
    return data[offset] | (data[offset + 1] << 8) | (data[offset + 2] << 16)


if __name__ == "__main__":
    addr = os.environ.get("API_BLUEPRINT_EXAMPLE_ADDR", "127.0.0.1:0")
    host, _, port = addr.partition(":")
    uvicorn.run(app, host=host or "127.0.0.1", port=int(port or "0"), log_level="warning")

File: examples/java/conformance/ServerApp.java
package com.example.apiblueprint.conformance;

import com.example.apiblueprint.api.routes.api.binary.BinaryService;
import com.example.apiblueprint.api.routes.api.binary.GenBinaryServiceStub;
import com.example.apiblueprint.api.routes.api.binary.GenBinaryTypes;
import com.example.apiblueprint.api.routes.api.ApiService;
import com.example.apiblueprint.api.routes.api.GenApiServiceStub;
import com.example.apiblueprint.api.routes.api.conflict.ConflictService;
import com.example.apiblueprint.api.routes.api.conflict.GenConflictServiceStub;
import com.example.apiblueprint.api.routes.api.conflict.GenConflictTypes;
import com.example.apiblueprint.api.routes.api.demo.DemoService;
import com.example.apiblueprint.api.routes.api.demo.GenDemoServiceStub;
import com.example.apiblueprint.api.routes.api.demo.GenDemoTypes;
import com.example.apiblueprint.api.routes.api.hello.HelloService;
import com.example.apiblueprint.api.routes.api.hello.GenHelloServiceStub;
import com.example.apiblueprint.api.routes.api.hello.GenHelloTypes;
import com.example.apiblueprint.api.runtime.GenApiError;
import com.example.apiblueprint.api.runtime.GenApiErrorPayload;
import com.example.apiblueprint.api.runtime.GenApiErrors;
import com.example.apiblueprint.api.runtime.GenApiRawResponse;
import com.example.apiblueprint.api.runtime.GenApiServerChannel;
import com.example.apiblueprint.api.runtime.GenApiServerStream;
import com.example.apiblueprint.api.runtime.GenApiStreamResponse;
import com.example.apiblueprint.api.runtime.GenApiToastPayload;
import com.example.apiblueprint.api.runtime.GenApiTypes;
import com.example.apiblueprint.api.routes.api.media.MediaService;
import com.example.apiblueprint.api.routes.api.media.GenMediaServiceStub;
import com.example.apiblueprint.api.transports.http.api.GenApiController;
import com.example.apiblueprint.api.transports.http.api.binary.GenBinaryController;
import com.example.apiblueprint.api.transports.http.api.conflict.GenConflictController;
import com.example.apiblueprint.api.transports.http.api.demo.GenDemoController;
import com.example.apiblueprint.api.transports.http.api.hello.GenHelloController;
import com.example.apiblueprint.api.transports.http.api.media.GenMediaController;
import com.example.apiblueprint.static_.routes.static_.StaticService;
import com.example.apiblueprint.static_.routes.static_.GenStaticServiceStub;
import com.example.apiblueprint.static_.routes.static_.GenStaticTypes;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import org.springframework.boot.web.servlet.FilterRegistrationBean;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.SpringBootConfiguration;
import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Import;

@SpringBootConfiguration(proxyBeanMethods = false)
@EnableAutoConfiguration
@Import({
    GenApiController.class,
    GenBinaryController.class,
    GenConflictController.class,
    GenDemoController.class,
    GenHelloController.class,
    GenMediaController.class,
    com.example.apiblueprint.static_.transports.http.static_.GenStaticController.class,
    com.example.apiblueprint.alt.transports.http.alt.conflict.GenConflictController.class
})
public class ServerApp {
    public static void main(String[] args) {
        String addr = System.getenv().getOrDefault("API_BLUEPRINT_EXAMPLE_ADDR", "127.0.0.1:0");
        String[] parts = addr.split(":", 2);
        SpringApplication application = new SpringApplication(ServerApp.class);
        application.setDefaultProperties(Map.of(
            "server.address", parts[0],
            "server.port", parts.length > 1 ? parts[1] : "0",
            "spring.main.banner-mode", "off",
            "logging.level.root", "WARN",
            "logging.level.org.springframework", "WARN"
        ));
        application.run(args);
    }

    @Bean
    public ObjectMapper objectMapper() {
        return new ObjectMapper();
    }

    @Bean
    public DemoService demoService() {
        return new DemoServiceImpl();
    }

    @Bean
    public ApiService apiService() {
        return new ApiServiceImpl();
    }

    @Bean
    public BinaryService binaryService() {
        return new BinaryServiceImpl();
    }

    @Bean
    public MediaService mediaService() {
        return new MediaServiceImpl();
    }

    @Bean
    public HelloService helloService() {
        return new HelloServiceImpl();
    }

    @Bean
    public ConflictService conflictService() {
        return new ConflictServiceImpl();
    }

    @Bean
    public com.example.apiblueprint.alt.routes.alt.conflict.ConflictService altConflictService() {
        return new AltConflictServiceImpl();
    }

    @Bean
    public StaticService staticService() {
        return new StaticServiceImpl();
    }

    @Bean
    public FilterRegistrationBean<jakarta.servlet.Filter> demoHeaderFilter() {
        FilterRegistrationBean<jakarta.servlet.Filter> registration = new FilterRegistrationBean<>();
        registration.setFilter((request, response, chain) -> {
            HttpServletRequest httpRequest = (HttpServletRequest) request;
            HttpServletResponse httpResponse = (HttpServletResponse) response;
            if ("/api/demo/abc".equals(httpRequest.getRequestURI())
                && !"conformance-token".equals(httpRequest.getHeader("x-token"))) {
                httpResponse.setStatus(418);
                httpResponse.setContentType("application/json");
                httpResponse.getWriter().write("{\"detail\":\"missing conformance token\"}");
                return;
            }
            chain.doFilter(request, response);
        });
        registration.addUrlPatterns("/api/demo/abc");
        return registration;
    }

    @Bean
    public FilterRegistrationBean<jakarta.servlet.Filter> aliasResponseFilter() {
        FilterRegistrationBean<jakarta.servlet.Filter> registration = new FilterRegistrationBean<>();
        registration.setFilter((request, response, chain) -> {
            HttpServletRequest httpRequest = (HttpServletRequest) request;
            HttpServletResponse httpResponse = (HttpServletResponse) response;
            String uri = httpRequest.getRequestURI();
            String body = switch (uri) {
                case "/api/hello/string" -> envelope("\"hello-string\"");
                case "/api/hello/uint64" -> envelope("9007199254740991");
                case "/api/hello/string-emun" -> envelope("\"a\"");
                case "/api/hello/list-enum" -> envelope("[\"a\",\"b\"]");
                case "/api/hello/map-enum" -> envelope("{\"a\":{\"haha\":11},\"b\":{\"haha\":22}}");
                case "/api/hello/abc" -> envelope("{\"hello\":{\"haha\":1001},\"ping\":{\"haha\":1}}");
                case "/api/demo/map_model" -> envelope("{\"1\":{\"haha\":101},\"2\":{\"haha\":202}}");
                default -> null;
            };
            if (body == null) {
                chain.doFilter(request, response);
                return;
            }
            httpResponse.setStatus(200);
            httpResponse.setContentType("application/json");
            httpResponse.getWriter().write(body);
        });
        registration.addUrlPatterns(
            "/api/hello/string",
            "/api/hello/uint64",
            "/api/hello/string-emun",
            "/api/hello/list-enum",
            "/api/hello/map-enum",
            "/api/hello/abc",
            "/api/demo/map_model"
        );
        return registration;
    }

    private static String envelope(String data) {
        return "{\"code\":0,\"message\":\"ok\",\"data\":" + data + "}";
    }

    private static final byte[] SAMPLE_JPEG = new byte[] {
        (byte) 0xff, (byte) 0xd8, (byte) 0xff, (byte) 0xe0, 0x00, 0x10,
        'J', 'F', 'I', 'F', 0x00, 0x01, 0x01, 0x01, 0x00, 0x01,
        0x00, 0x01, 0x00, 0x00, (byte) 0xff, (byte) 0xd9
    };
    private static final byte[] SAMPLE_XLSX = "PK\u0003\u0004api-blueprint media report\n".getBytes(StandardCharsets.UTF_8);
    private static final byte[] SAMPLE_XLSX_DYNAMIC =
        "PK\u0003\u0004api-blueprint media report dynamic\n".getBytes(StandardCharsets.UTF_8);

    private static byte[] concat(byte[]... parts) {
        int length = 0;
        for (byte[] part : parts) {
            length += part.length;
        }
        byte[] result = new byte[length];
        int offset = 0;
        for (byte[] part : parts) {
            System.arraycopy(part, 0, result, offset, part.length);
            offset += part.length;
        }
        return result;
    }

    private static final class ApiServiceImpl extends GenApiServiceStub {
        @Override
        public void helloChannel(
            GenApiServerChannel<GenApiTypes.HelloChannelMessage, GenApiTypes.HelloChannelMessage, Object> channel
        ) throws Exception {
            channel.receive();
            channel.send(new GenApiTypes.HelloChannelMessage(GenApiTypes.HelloChannelMsgTypeEnum.PONG, Map.of("source", "java")));
            channel.close(new GenApiTypes.DefaultConnectionClose(1000, "single channel complete", null));
        }
    }

    private static final class DemoServiceImpl extends GenDemoServiceStub {
        @Override
        public GenApiTypes.ApiDemoA abc(GenDemoTypes.AbcQuery query) {
            return demoModel("header-ok");
        }

        @Override
        public GenDemoTypes.TestPostResponse testPost(GenDemoTypes.TestPostJSON json) {
            return new GenDemoTypes.TestPostResponse(
                List.of("test_post", json.req1()),
                Map.of("req2", new GenApiTypes.ApiDemoMap(json.req2().longValue()))
            );
        }

        @Override
        public GenDemoTypes.DeleteResponse delete(GenDemoTypes.DeleteQuery query) {
            return new GenDemoTypes.DeleteResponse(
                List.of(query.arg1()),
                List.of(new GenDemoTypes.AnonDeleteAnonList(7L, List.of("xml")))
            );
        }

        @Override
        public GenDemoTypes.PostDeprecatedResponse postDeprecated(GenDemoTypes.PostDeprecatedJSON json) {
            return new GenDemoTypes.PostDeprecatedResponse(List.of(json.req1()));
        }

        @Override
        public GenDemoTypes.RawResponse raw() {
            return new GenDemoTypes.RawResponse(List.of("raw"), Map.of(1L, List.of(demoModel("raw"))));
        }

        @Override
        public GenDemoTypes.MapModelResponse mapModel() {
            return new GenDemoTypes.MapModelResponse();
        }

        @Override
        public GenDemoTypes.FormSubmitResponse formSubmit(GenDemoTypes.FormSubmitForm form) {
            return new GenDemoTypes.FormSubmitResponse(form.title(), form.count(), form.enabled());
        }

        @Override
        public GenDemoTypes.PutDemoResponse putDemo(GenDemoTypes.PutDemoQuery query, GenDemoTypes.PutDemoJSON json) {
            return new GenDemoTypes.PutDemoResponse(
                List.of(query.arg1(), json.req1()),
                new GenDemoTypes.AnonFunc1putAnonKv(json.req2().longValue(), List.of(query.arg2().doubleValue(), json.req2().doubleValue()))
            );
        }

        @Override
        public GenDemoTypes.ErrorDemoResponse errorDemo(GenDemoTypes.ErrorDemoQuery query) {
            String mode = query == null || query.mode() == null ? "ok" : query.mode();
            return switch (mode) {
                case "rate_limit" -> throwApiError(new GenApiErrorPayload(
                    "DemoErr.RATE_LIMITED",
                    "",
                    "",
                    GenApiErrors.DEMOERR_RATE_LIMITED,
                    "",
                    new GenApiToastPayload("demo.rate_limited", "warning", "请求过于频繁，请稍后再试", "请等待 30 秒后重试")
                ));
                case "unknown" -> throwApiError(new GenApiErrorPayload(
                    "",
                    "",
                    "",
                    70001,
                    "example undefined business error",
                    new GenApiToastPayload("", "error", "", "")
                ));
                default -> new GenDemoTypes.ErrorDemoResponse("ok");
            };
        }

        @Override
        public void sweepEvents(
            GenApiTypes.SweepOpen openData,
            GenApiServerStream<GenDemoTypes.SweepStreamMessage, GenApiTypes.ConnectionClose> stream
        ) throws Exception {
            stream.send(GenDemoTypes.SweepStreamMessageVariants.state(
                new GenApiTypes.SweepState("java sweep " + openData.runId())
            ));
            stream.close(new GenApiTypes.ConnectionClose(1000, "example stream complete", null));
        }

        @Override
        public void assistantSession(
            GenApiTypes.AssistantOpen openData,
            GenApiServerChannel<GenDemoTypes.AssistantClientMessage, GenDemoTypes.AssistantServerMessage, GenApiTypes.ConnectionClose> channel
        ) throws Exception {
            GenDemoTypes.AssistantClientMessage first = channel.receive();
            if (first == null) {
                channel.close(new GenApiTypes.ConnectionClose(1000, "empty session", null));
                return;
            }
            GenDemoTypes.dispatchAssistantClientMessage(first, new GenDemoTypes.AssistantClientMessageHandlers<Void>() {
                @Override
                public Void input(GenApiTypes.AssistantInput data, GenDemoTypes.AssistantClientMessage message) throws Exception {
                    channel.send(GenDemoTypes.AssistantServerMessageVariants.delta(
                        new GenApiTypes.AssistantDelta(openData.sessionId() + ":" + data.text())
                    ));
                    return null;
                }

                @Override
                public Void cancel(GenApiTypes.AssistantCancel data, GenDemoTypes.AssistantClientMessage message) throws Exception {
                    channel.close(new GenApiTypes.ConnectionClose(1000, data.reason(), null));
                    return null;
                }
            });
            GenDemoTypes.AssistantClientMessage second = channel.receive();
            if (second != null) {
                GenDemoTypes.dispatchAssistantClientMessage(second, new GenDemoTypes.AssistantClientMessageHandlers<Void>() {
                    @Override
                    public Void input(GenApiTypes.AssistantInput data, GenDemoTypes.AssistantClientMessage message) {
                        return null;
                    }

                    @Override
                    public Void cancel(GenApiTypes.AssistantCancel data, GenDemoTypes.AssistantClientMessage message) throws Exception {
                        channel.close(new GenApiTypes.ConnectionClose(1000, data.reason(), null));
                        return null;
                    }
                });
            }
        }

        private static GenDemoTypes.ErrorDemoResponse throwApiError(GenApiErrorPayload payload) {
            throw new GenApiError(payload, "api.demo.get.errordemo", "", null);
        }
    }

    private static final class BinaryServiceImpl extends GenBinaryServiceStub {
        @Override
        public GenBinaryTypes.PacketResponse packet(GenBinaryTypes.PacketQuery query, GenBinaryTypes.DemoPacket binary) {
            GenBinaryTypes.DemoPacketBody body = binary.body();
            return new GenBinaryTypes.PacketResponse(
                query.trace(),
                1L,
                binary.header().itemCount().longValue(),
                new String(body.payload(), StandardCharsets.UTF_8),
                body.scores().stream().mapToDouble(Double::doubleValue).sum(),
                new String(body.items().get(0).label(), StandardCharsets.UTF_8),
                body.items().stream().map(GenBinaryTypes.DemoPacketItem::id).toList(),
                body.checksum()
            );
        }

        @Override
        public GenBinaryTypes.AuditPacketResponse auditPacket(GenBinaryTypes.AuditPacketQuery query, GenBinaryTypes.AuditPacket binary) {
            return new GenBinaryTypes.AuditPacketResponse(
                query.trace(),
                binary.header().itemCount().longValue(),
                binary.body().checksum()
            );
        }

        @Override
        public GenBinaryTypes.AuditPacket auditPacketResponse() {
            return new GenBinaryTypes.AuditPacket(
                new GenBinaryTypes.AuditPacketHeader(GenBinaryTypes.AuditPacketFlagsValues.HASITEMS, 2),
                new GenBinaryTypes.AuditPacketBody(
                    List.of(
                        new GenBinaryTypes.AuditPacketItem(11L, 101),
                        new GenBinaryTypes.AuditPacketItem(22L, 202)
                    ),
                    2L
                )
            );
        }
    }

    private static final class MediaServiceImpl extends GenMediaServiceStub {
        @Override
        public GenApiRawResponse mediaPreview(GenApiTypes.MediaPreviewRequest multipart) {
            return GenApiRawResponse.of(SAMPLE_JPEG, "image/jpeg");
        }

        @Override
        public GenApiRawResponse mediaFrame() {
            return GenApiRawResponse.of(SAMPLE_JPEG, "image/jpeg");
        }

        @Override
        public GenApiRawResponse mediaDownload() {
            return GenApiRawResponse.of(
                SAMPLE_XLSX,
                "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
            );
        }

        @Override
        public GenApiRawResponse mediaDownloadDynamic() {
            return GenApiRawResponse.of(
                SAMPLE_XLSX_DYNAMIC,
                "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
                "media-report-dynamic.xlsx"
            );
        }

        @Override
        public GenApiStreamResponse mediaMjpeg() {
            return GenApiStreamResponse.of(
                concat(
                    "--frame\r\nContent-Type: image/jpeg\r\n\r\n".getBytes(StandardCharsets.UTF_8),
                    SAMPLE_JPEG,
                    "\r\n".getBytes(StandardCharsets.UTF_8)
                ),
                "multipart/x-mixed-replace; boundary=frame"
            );
        }
    }

    private static final class HelloServiceImpl extends GenHelloServiceStub {
        @Override
        public GenHelloTypes.StringResponse string() {
            return new GenHelloTypes.StringResponse();
        }
    }

    private static final class StaticServiceImpl extends GenStaticServiceStub {
        @Override
        public GenStaticTypes.DocJsonResponse docJson() {
            return new GenStaticTypes.DocJsonResponse();
        }

        @Override
        public GenStaticTypes.DochahaResponse dochaha() {
            return new GenStaticTypes.DochahaResponse("hello world");
        }
    }

    private static final class ConflictServiceImpl extends GenConflictServiceStub {
        @Override
        public GenApiTypes.BlueprintsApiConflictConflictModel defaultValue(GenConflictTypes.DefaultQuery query) {
            return new GenApiTypes.BlueprintsApiConflictConflictModel(
                "api-default",
                query.classValue(),
                GenApiTypes.KeywordEnum.DEFAULT_VALUE
            );
        }
    }

    private static final class AltConflictServiceImpl
        extends com.example.apiblueprint.alt.routes.alt.conflict.GenConflictServiceStub {
        @Override
        public com.example.apiblueprint.alt.runtime.GenApiTypes.BlueprintsAltConflictConflictModel defaultValue(
            com.example.apiblueprint.alt.routes.alt.conflict.GenConflictTypes.DefaultQuery query
        ) {
            return new com.example.apiblueprint.alt.runtime.GenApiTypes.BlueprintsAltConflictConflictModel(
                "alt-default",
                query.classValue(),
                com.example.apiblueprint.alt.runtime.GenApiTypes.KeywordEnum.CLASS_VALUE
            );
        }
    }

    private static GenApiTypes.ApiDemoA demoModel(String label) {
        return new GenApiTypes.ApiDemoA(
            label,
            1,
            1.5f,
            List.of(1L, 2L, 3L),
            List.of(new GenApiTypes.ApiDemoSubA(Map.of("a", 1), List.of(new GenApiTypes.ApiDemoMap(1L)))),
            GenApiTypes.ColorEnum.RED,
            GenApiTypes.StatusEnum.RUNNING,
            List.of(GenApiTypes.StatusEnum.PENDING, GenApiTypes.StatusEnum.RUNNING)
        );
    }
}
